### PR TITLE
flamenco: bpf loader rewrite

### PR DESCRIPTION
**Pull request by ibhatt-jumptrading**
Rewrite of the bpf loader

### CODE REVIEW

File: src/flamenco/runtime/program/fd_bpf_loader.c
[the remark this answers is not in the document]
Comment (ibhatt-jumptrading) [reply src/flamenco/runtime/program/fd_bpf_loader.c:1]: because its wrong

[the remark this answers is not in the document]
Comment (ibhatt-jumptrading) [reply src/flamenco/runtime/program/fd_bpf_loader.c:1]: following the naming convention provided by agave client

File: src/flamenco/runtime/program/fd_bpf_loader.c [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+  };
+
+  ulong validate_result = fd_vm_context_validate( &vm_ctx );
+  if( validate_result != FD_VM_SBPF_VALIDATE_SUCCESS ) {
+    FD_LOG_ERR(( "fd_vm_context_validate() failed: %lu", validate_result ));
+  }
+
+  fd_valloc_free( instr_ctx->valloc, fd_sbpf_program_delete( prog ) );
+  fd_valloc_free( instr_ctx->valloc, fd_sbpf_syscalls_delete( syscalls ) );
+  fd_valloc_free( instr_ctx->valloc, rodata );
```
[the remark this answers is not in the document]
Comment (ibhatt-jumptrading) [reply src/flamenco/runtime/program/fd_bpf_loader.c]: Yep this was just from copying over from the old implementation 🙂

File: src/flamenco/runtime/program/fd_bpf_loader_v3_program.h [hunk 2]
```diff
@@ -39,11 +26,20 @@ int
...
+/* fd_bpf_loader_v3_program_execute is the shared entry point for bpf 
+   user-defined programs as well as executions of the program itself. */
+
+int
+fd_bpf_loader_v3_program_execute( fd_exec_instr_ctx_t instr_ctx );
+
+/* TODO: add comment here */
+
 fd_account_meta_t const *
 read_bpf_upgradeable_loader_state_for_program( fd_exec_txn_ctx_t * txn_ctx,
```
Comment (ibhatt-jumptrading) [inline src/flamenco/runtime/program/fd_bpf_loader_v3_program.h:38]: This needs to get refactored and cleaned up, but leaving for now as it's used in other parts of the codebase. It serves the same funciton as `read_bpf_upgradeable_loader_state_for_program`

**Review by topointon-jump — COMMENTED** [review]
Mid-way through reviewing, will finish by EOD. Feel free to ignore these comments or address in a follow-up PR :)

**Review by topointon-jump — CHANGES_REQUESTED** [review]
The sysvar account checks are consensus-critical, do you mind addressing before merging? 🙏

File: src/flamenco/runtime/program/fd_bpf_loader_v3_program.c [hunk 3]
```diff
@@ -71,117 +94,234 @@ read_bpf_upgradeable_loader_state_for_program( fd_exec_txn_ctx_t * txn_ctx,
...
 
-int write_bpf_upgradeable_loader_state( fd_exec_instr_ctx_t * instr_ctx, fd_pubkey_t const * program_acc, fd_bpf_upgradeable_loader_state_t const * loader_state) {
-  int err = 0;
-  ulong encoded_loader_state_size = fd_bpf_upgradeable_loader_state_size( loader_state );
+/* https://github.com/anza-xyz/agave/blob/574bae8fefc0ed256b55340b9d87b7689bcdf222/programs/bpf_loader/src/lib.rs#L105-L171 */
+int
+deploy_program( fd_exec_instr_ctx_t * instr_ctx,
+                uchar * const         programdata,
+                ulong                 programdata_size ) {   
+  FD_SCRATCH_SCOPE_BEGIN {
```
Comment (topointon-jump) [inline src/flamenco/runtime/program/fd_bpf_loader_v3_program.c:111]: You could also just have a single `FD_SCRATCH_SCOPE_[BEGIN,END]` in `fd_bpf_loader_v3_program_execute`, which might be cleaner.

File: src/flamenco/runtime/program/fd_bpf_loader_v3_program.c [hunk 4]
```diff
@@ -190,1352 +330,1149 @@ fd_bpf_loader_v3_user_execute( fd_exec_instr_ctx_t ctx ) {
...
+      fd_native_cpi_create_account_meta( programdata_key, 1U, 1U, &acct_metas[ 1UL ] );
+      fd_native_cpi_create_account_meta( buffer_key,      0U, 1U, &acct_metas[ 2UL ] );
+      fd_pubkey_t signers[ 2UL ];
+      ulong signers_cnt = 2UL;
+      signers[ 0UL ] = *payer_key;
+      signers[ 1UL ] = *programdata_key;
+
+      err = fd_native_cpi_execute_system_program_instruction( instr_ctx, &instr, acct_metas, 
+                                                              3UL, signers, signers_cnt );
+      if( err ) {
```
Comment (topointon-jump) [inline src/flamenco/runtime/program/fd_bpf_loader_v3_program.c:647]:
```suggestion
      if( FD_UNLIKELY( err ) ) {
```

Do you mind adding this hint to all the conditionals as appropriate?

File: src/flamenco/runtime/program/fd_bpf_loader_v3_program.c [hunk 5]
```diff
@@ -190,1352 +330,1149 @@ fd_bpf_loader_v3_user_execute( fd_exec_instr_ctx_t ctx ) {
...
+      fd_vm_rust_account_meta_t * acct_metas = (fd_vm_rust_account_meta_t*) 
+                                               fd_scratch_alloc( FD_VM_RUST_ACCOUNT_META_ALIGN, 
+                                                                 3UL * sizeof(fd_vm_rust_account_meta_t) );
+      fd_native_cpi_create_account_meta( payer_key,       1U, 1U, &acct_metas[ 0UL ] ); 
+      fd_native_cpi_create_account_meta( programdata_key, 1U, 1U, &acct_metas[ 1UL ] );
+      fd_native_cpi_create_account_meta( buffer_key,      0U, 1U, &acct_metas[ 2UL ] );
+      fd_pubkey_t signers[ 2UL ];
+      ulong signers_cnt = 2UL;
+      signers[ 0UL ] = *payer_key;
+      signers[ 1UL ] = *programdata_key;
```
Comment (topointon-jump) [inline src/flamenco/runtime/program/fd_bpf_loader_v3_program.c:643]: These signers look wrong - Agave uses these keys as seeds to derive signer PDA's from, not the keys themselves: https://github.com/anza-xyz/agave/blob/574bae8fefc0ed256b55340b9d87b7689bcdf222/programs/bpf_loader/src/lib.rs#L638-L641

File: src/flamenco/runtime/program/fd_bpf_loader_v3_program.c [hunk 6]
```diff
@@ -190,1352 +330,1149 @@ fd_bpf_loader_v3_user_execute( fd_exec_instr_ctx_t ctx ) {
...
-    meta->info.executable = 0;
-    fd_memcpy(&meta->info.owner, fd_solana_bpf_loader_upgradeable_program_id.key, sizeof(fd_pubkey_t));
-    meta->info.lamports = fd_rent_exempt_minimum_balance(ctx.slot_ctx, meta->dlen);
-    meta->info.rent_epoch = 0;
-
-    buffer_data_len = fd_ulong_sat_sub(buffer_rec->meta->dlen, BUFFER_METADATA_SIZE);
+    if( buffer_data_offset>=buffer->meta->dlen ) {
+      FD_LOG_WARNING(( "Buffer data offset is out of bounds" ));
+      return FD_EXECUTOR_INSTR_ERR_ACC_DATA_TOO_SMALL;
+    }
```
Comment (topointon-jump) [inline src/flamenco/runtime/program/fd_bpf_loader_v3_program.c:663]: Where are these checks in Agave? These don't line up with the source - do you mind adding a permalink 🙏

File: src/flamenco/runtime/program/fd_bpf_loader_v3_program.c [hunk 7]
```diff
@@ -190,1352 +330,1149 @@ fd_bpf_loader_v3_user_execute( fd_exec_instr_ctx_t ctx ) {
...
+    uchar * dst_slice       = programdata->data + PROGRAMDATA_METADATA_SIZE;
+    ulong dst_slice_len     = buffer_data_len; 
+    const uchar * src_slice = buffer->const_data + buffer_data_offset;
+    fd_memcpy( dst_slice, src_slice, dst_slice_len );
+
+    /* Update the programdata's metadata */
+    programdata->meta->dlen            = programdata_len;
+    programdata->meta->info.executable = 0U;
+    programdata->meta->info.lamports   = fd_rent_exempt_minimum_balance( instr_ctx->slot_ctx, programdata_len );
+    programdata->meta->info.rent_epoch = 0UL;
```
Comment (topointon-jump) [inline src/flamenco/runtime/program/fd_bpf_loader_v3_program.c:713]: Where are these checks in Agave? These don't line up with the source - do you mind adding a permalink 🙏

**Review by topointon-jump — CHANGES_REQUESTED** [review]
Overall looks great! Spotted a few consensus-critical things

File: src/flamenco/runtime/program/fd_bpf_loader_v3_program.c [hunk 8]
```diff
@@ -190,1352 +330,1149 @@ fd_bpf_loader_v3_user_execute( fd_exec_instr_ctx_t ctx ) {
...
+      FD_LOG_WARNING(( "Bincode decode for account failed" ));
       return err;
     }
 
-    if ( fd_bpf_upgradeable_loader_state_is_buffer( &loader_state ) ) {
-      if ( new_authority_acc==NULL ) {
-        return FD_EXECUTOR_INSTR_ERR_INCORRECT_AUTHORITY;
+    if( fd_bpf_upgradeable_loader_state_is_buffer( &account_state ) ) {
+      if( !new_authority ) {
+        FD_LOG_WARNING(( "Buffer authority is not optional" ));
```
Comment (topointon-jump) [inline src/flamenco/runtime/program/fd_bpf_loader_v3_program.c:971]:
```suggestion
        FD_LOG_WARNING(( "Buffer authority is not optional" ));
        return FD_EXECUTOR_INSTR_ERR_INCORRECT_AUTHORITY;
```

File: src/flamenco/runtime/program/fd_bpf_loader_v3_program.c [hunk 9]
```diff
@@ -190,1352 +330,1149 @@ fd_bpf_loader_v3_user_execute( fd_exec_instr_ctx_t ctx ) {
...
-
-      if (memcmp(&program_acc_state.inner.program.programdata_address, close_acc, sizeof(fd_pubkey_t)) != 0) {
-        return FD_EXECUTOR_INSTR_ERR_INVALID_ARG;
-      }
-
-      fd_pubkey_t const * authority_acc = &txn_accs[instr_acc_idxs[2]];
-
-      err = common_close_account(ctx, authority_acc, close_acc_rec->meta, recipient_acc_rec->meta, &loader_state, close_acc);
-      if (err != 0) {
+      account_state.inner.program_data.upgrade_authority_address = new_authority;
```
Comment (topointon-jump) [inline src/flamenco/runtime/program/fd_bpf_loader_v3_program.c:1004]:
```suggestion
      account_state.inner.program_data.upgrade_authority_address = new_authority;
      account_state.inner.program_data.slot = clock.slot;
```

File: src/flamenco/runtime/program/fd_bpf_loader_v3_program.c [hunk 10]
```diff
@@ -190,1352 +330,1149 @@ fd_bpf_loader_v3_user_execute( fd_exec_instr_ctx_t ctx ) {
...
-      if ( !fd_instr_acc_is_signer_idx( ctx.instr, 2 ) ) {
+      if( !fd_instr_acc_is_signer_idx( instr_ctx->instr, 2UL ) ) {
+        FD_LOG_WARNING(( "New authority did not sign" ));
         return FD_EXECUTOR_INSTR_ERR_MISSING_REQUIRED_SIGNATURE;
       }
-
-      *loader_state.inner.program_data.upgrade_authority_address = *new_authority_acc;
-
-      return write_bpf_upgradeable_loader_state( &ctx, loader_acc, &loader_state );
+      account_state.inner.program_data.upgrade_authority_address = (fd_pubkey_t*)new_authority_key;
```
Comment (topointon-jump) [inline src/flamenco/runtime/program/fd_bpf_loader_v3_program.c:1076]:
```suggestion
      account_state.inner.program_data.upgrade_authority_address = (fd_pubkey_t*)new_authority_key;
      account_state.inner.program_data.slot = clock.slot;
```

File: src/flamenco/runtime/program/fd_bpf_loader_v3_program.c [hunk 11]
```diff
@@ -190,1352 +330,1149 @@ fd_bpf_loader_v3_user_execute( fd_exec_instr_ctx_t ctx ) {
...
-        //     program_key,
-        //     Arc::new(LoadedProgram::new_tombstone(
-        //         clock.slot,
-        //         LoadedProgramType::Closed,
-        //     )),
-        // );
+        /* The Agave client updates the account state upon closing an account
+           in their loaded prograam cache. Checking for a program can be
+           checked by checking to see if the programdata account's loader state 
+           is unitialized. */
```
Comment (topointon-jump) [inline src/flamenco/runtime/program/fd_bpf_loader_v3_program.c:1193]: Agave delays the visibility of program deployments: whenever a program is un-/re-/deployed, it becomes unavailable for the rest of the transaction. They do this by modifying the cache entry. Do we have these same semantics?

(See https://github.com/solana-labs/solana/pull/29654 and `delay_visibility_of_program_deployment`)

**Review by topointon-jump — CHANGES_REQUESTED** [review]
Found a few more consensus-critical issues - almost there 😄

File: src/flamenco/runtime/program/fd_bpf_loader_v3_program.c (hunk 11 above)
Comment (topointon-jump) [reply src/flamenco/runtime/program/fd_bpf_loader_v3_program.c:1193]: We will add ledger tests for this

File: src/flamenco/runtime/program/fd_bpf_loader_v3_program.c [hunk 12]
```diff
@@ -71,1471 +94,1399 @@ read_bpf_upgradeable_loader_state_for_program( fd_exec_txn_ctx_t * txn_ctx,
...
+        &instr,
+        acct_metas,
+        2UL,
+        signers,
+        1UL
+      );
+      if( FD_UNLIKELY( err ) ) {
+        return err;
+      }
+    }
```
Comment (topointon-jump) [inline src/flamenco/runtime/program/fd_bpf_loader_v3_program.c:1335]: Agave doesn't pass any signers: https://github.com/anza-xyz/agave/blob/574bae8fefc0ed256b55340b9d87b7689bcdf222/programs/bpf_loader/src/lib.rs#L1250-L1254

**Review by topointon-jump — CHANGES_REQUESTED** [review]
Last consensus-critical bug 🎉

File: src/flamenco/runtime/program/fd_bpf_loader_v3_program.c (hunk 12 above)
Comment (topointon-jump) [reply src/flamenco/runtime/program/fd_bpf_loader_v3_program.c:1335]: Looks like GitHub put this comment in the wrong place - I was referring to https://github.com/firedancer-io/firedancer/blob/00bf781b080b7a37ac82bf4b588048d31ba820bb/src/flamenco/runtime/program/fd_bpf_loader_v3_program.c#L1321-L1322

File: src/flamenco/runtime/program/fd_bpf_loader_v3_program.c [hunk 13]
```diff
@@ -190,1352 +330,1149 @@ fd_bpf_loader_v3_user_execute( fd_exec_instr_ctx_t ctx ) {
...
-    // FIXME: Do checked addition
-    ulong buffer_lamports = buffer_rec->meta->info.lamports;
-    payer->meta->info.lamports += buffer_lamports;
-    // TODO: Does this mean this account is dead?
-    buffer_rec->meta->info.lamports  = 0;
-
-    // TODO: deploy program
-    err = setup_program( &ctx, buffer_rec->data + BUFFER_METADATA_SIZE, buffer_data_len );
-    if ( err != 0 ) {
+    err = fd_instr_borrowed_account_modify_idx( instr_ctx, 0UL, 0UL, &payer );
```
Comment (topointon-jump) [inline src/flamenco/runtime/program/fd_bpf_loader_v3_program.c:609]: We should use `fd_borrowed_account_acquire_write`/`fd_borrowed_account_acquire_read`/`fd_borrowed_account_release_write`/`fd_borrowed_account_release_read` like we do in the other native programs.

We can move to the scoping macro in a follow-up PR.

File: src/flamenco/runtime/fd_pubkey_utils.c [hunk 14]
```diff
@@ -31,3 +33,72 @@ fd_pubkey_create_with_seed( fd_exec_instr_ctx_t const * ctx,
...
+    
+    fd_sha256_append( &sha, seed, 32UL );
+    if( bump_seed ) {
+      fd_sha256_append( &sha, bump_seed, 1UL );
+    }
+    fd_sha256_append( &sha, program_id,              sizeof(fd_pubkey_t) );
+    fd_sha256_append( &sha, "ProgramDerivedAddress", 21UL                );
+
+    fd_sha256_fini( &sha, out );
+  }
```
Comment (topointon-jump) [inline src/flamenco/runtime/fd_pubkey_utils.c:67]: This iteration is wrong - we should hash all the seeds, then hash the bump seed, then hash the program id, then hash the marker. This loop hashes the bump seed, program id and marker every time.

See https://github.com/firedancer-io/firedancer/blob/vm-belt-sanding-public/src/flamenco/vm/syscall/fd_vm_syscall_pda.c#L12-L69

**Review by topointon-jump — CHANGES_REQUESTED** [review]
Last round of review! The PDA bug should definitely be fixed before merging 🙏

File: src/flamenco/runtime/program/fd_bpf_loader_v3_program.c [hunk 15]
```diff
@@ -71,1471 +97,1385 @@ read_bpf_upgradeable_loader_state_for_program( fd_exec_txn_ctx_t * txn_ctx,
...
-        if ( err ) {
-          return err;
+/* process_instruction_inner() */
+/* https://github.com/anza-xyz/agave/blob/77daab497df191ef485a7ad36ed291c1874596e5/programs/bpf_loader/src/lib.rs#L394-L564 */
+int 
+fd_bpf_loader_v3_program_execute( fd_exec_instr_ctx_t ctx ) {
+  FD_SCRATCH_SCOPE_BEGIN {
+    /* https://github.com/anza-xyz/agave/blob/77daab497df191ef485a7ad36ed291c1874596e5/programs/bpf_loader/src/lib.rs#L491-L529 */
+    fd_borrowed_account_t * program_account = NULL;
+    fd_pubkey_t const *     program_id      = &ctx.instr->program_id_pubkey;
```
[the remark this answers is not in the document]
Comment (ibhatt-jumptrading) [reply src/flamenco/runtime/program/fd_bpf_loader_v3_program.c:1376]: committing with the same version that matches the rest of my commits

**Review by topointon-jump — DISMISSED** [review]
🎉

**Pull request merged**